Split the playfield columns and change the key number by mod

andy840119 opened the pull request:
Closes #716.
.
1. Split the Playfield columns by ManiaModKeyCoop
2. Can change the key number by ManiaKeyMod

andy840119 committed:
https://github.com/ppy/osu/issues/716

1. split the playfield columns by ManiaModKeyCoop
2. can chaneg the key number by ManiaKeyMod

andy840119 committed:
Fixed the alert from AppVeyor.
maybe.

andy840119 committed:
after resharper : (
.
裝上Resharper，從此VS的速度一片黑暗

andy840119 committed:
pray

andy840119 committed:
Appveyor Chan, Please

andy840119 committed:
Merge branch 'master' into master

smoogipoo reviewed (changes requested):
I had a pass through this and while in general it needs a lot of code cleanup, I think the biggest issue is that it's not as extensible as I want it to be. I don't want to stop at Co-op mode, but this kinda forces us down that path.

I've started cleaning up the mania codebase to be more extensible for this type of stuff: https://github.com/ppy/osu/pull/1845/files please re-base off of that and change the code as required. You will need to make `ManiaPlayfield` accept a `List<StageDefinition>`, and each of those will contain the number of columns for the stages.

Please rename all "group" to "stage", it's a bit more descriptive of what it represents. Avoid "single" in your foreach iterators - instead use "stage".

andy840119 committed:
Merge branch 'master' into master

andy840119 committed:
Merge branch 'mania-rc-cleanup' of https://github.com/smoogipoo/osu

# Conflicts:
#	osu.Game.Rulesets.Mania/UI/ManiaRulesetContainer.cs

andy840119 commented:
OK I'll check this problem and modified it  : )

andy840119 committed:
Merge branch 'master' of https://github.com/andy840119/osu

andy840119 committed:
fix some error that smoogipoo says

andy840119 committed:
group -> stage

andy840119 committed:
1. fix TestCaseManiaHitObjects broken
2. add (4+4) (2+4+2) (1+8+1) column stages step in TestCaseManiaPlayfield

andy840119 committed:
White space needs to be trimmed

andy840119 committed:
Merge branch 'mania-rc-cleanup' of https://github.com/smoogipoo/osu into mania_fix

# Conflicts:
#	osu-framework

andy840119 committed:
fix framework branch

andy840119 committed:
Merge branch 'master' of https://github.com/andy840119/osu

andy840119 committed:
update framework again

andy840119 committed:
fix Empty argument list is redundant

andy840119 committed:
miss

andy840119 committed:
Merge branch 'master' into master

andy840119 committed:
1. update from ppy master
2. update header

andy840119 committed:
update resource branch

andy840119 committed:
update framework again....

andy840119 committed:
Merge branch 'master' into master

andy840119 committed:
Merge branch 'master' into master

andy840119 committed:
Merge branch 'master' into master

Aergwyn commented on osu.Game.Rulesets.Mania/UI/ManiaPlayfield.cs:
```diff
@@ -1,153 +1,103 @@
+                var list = new List<Column>();
+                foreach (var stage in listColumnStages)
+                {
+                    list.AddRange(stage.Columns);
```
Could this be simplified to `get => listColumnStages.Select(c => c.Columns)`?

andy840119 replied:
I tried it and it should use 
public List<Column> Columns => listColumnStages.SelectMany(x => x.Columns).ToList();
instead

andy840119 committed:
use public List<Column> Columns => listColumnStages.SelectMany(x => x.Columns).ToList(); instead

andy840119 committed:
Merge branch 'master' of https://github.com/ppy/osu

# Conflicts:
#	osu.Game.Rulesets.Mania/Mods/ManiaModGravity.cs

andy840119 committed:
Merge branch 'master' into master

andy840119 committed:
Merge branch 'master' into master

andy840119 committed:
1. because set stage's width doesn't help, so use margin instead
2. adjust spacing when has different number of stages

andy840119 committed:
fix CI error

andy840119 committed:
Merge branch 'master' into master

smoogipoo commented on osu.Game.Rulesets.Mania/UI/ManiaPlayfield.cs:
```diff
@@ -157,82 +99,45 @@ public ManiaPlayfield(int columnCount)
-        /// <param name="column">The 0-based column index.</param>
-        /// <returns>Whether the column is a special column.</returns>
-        private bool isSpecialColumn(int column)
+        public override void Add(DrawableHitObject h) => Columns.ElementAt(((ManiaHitObject)h.HitObject).Column).Add(h);
```
This (along with OnJudgement above) should be done in the following three steps:

1. Make this call `Add` on the stage which the column belongs to, with the hitobject. You can use the `getStageByColumn` method here.
2. Have the stage's OnJudgement do the `AddJudgement` stuff.
3. Remove the OnJudgement call for the column, that should be handled by the column itself.

smoogipoo commented on osu.Game.Rulesets.Mania/UI/ManiaPlayfield.cs:
```diff
@@ -157,82 +99,45 @@ public ManiaPlayfield(int columnCount)
-                Color4 colour = normalColumnColours[i % normalColumnColours.Count];
-                nonSpecialColumns[i].AccentColour = colour;
-                nonSpecialColumns[nonSpecialColumns.Count - 1 - i].AccentColour = colour;
+                single.Judgements.Scale = Scale;
```
How about giving the stages an `Inverted` BindableBool, and having them decide their scale?

smoogipoo commented on osu.Game.Rulesets.Mania/UI/ManiaPlayfield.cs:
```diff
@@ -1,153 +1,95 @@
+                {
+                    var c = new Column
+                    {
+                        //c.Action = c.IsSpecial ? ManiaAction.Special : currentAction++;
```
This isn't going to work - special columns are broken right now. Mania's Key1 starts at value = 10 to allow for, e.g. `ManiaAction.Special + 1` for the second special column.

andy840119 replied:
or just cancel ManiaAction.Special?
Because in co-op mode need two Special in ManiaAction,
and user need to reset the setting when SpecialColumnPosition is left or right.

smoogipoo replied:
Not sure what you mean by cancel? Each stage should get a special column by doing something like `ManiaAction.Special + stageIndex`

andy840119 committed:
Merge branch 'master' into master

andy840119 reviewed (commented):
\change/

smoogipoo commented:
They KeyCoop mod doesn't look to be implemented yet. I added a description to my previous cleanup PR for how this can be done (https://github.com/ppy/osu/pull/1845).

andy840119 committed:
update rom ppy master
.
Find out that has some problem long hit note will not show hit result

andy840119 committed:
fix :
1. Add space before {, before Duration, and after 1000
2. How about just ManiaStage?
3. This is really just Columns.Count, you're not saving much with this extra variable here.

andy840119 committed:
fix the problem that smoogipoo suggest
1. Rename to stages
2. This should be checking for null or 0, with two exceptions:
3. Space before and after =.
4. foreach (var stage ...
5. Rename to getStageByColumn, and actualColumn -> column.

andy840119 committed:
Fix :
This (along with OnJudgement above) should be done in the following three steps:
.
i'm not sure does it means the code i write

andy840119 committed:
Fix :
1. This (along with OnJudgement above) should be done in the following three steps:
2. How about giving the stages an Inverted BindableBool, and having them decide their scale?

andy840119 committed:
This should be done in the stage instead of the playfield.

andy840119 committed:
fix CI error

andy840119 commented:
remain
1. adjust ManiaStage's size
2. ManiaAction and special key issue
3. use FillFlowContainer or GrodContainer to contain ManiaStage
.
And due to DrawableHitObject.AddNested() changed
maina hold note cannot trigger OnJudgement()

smoogipoo commented:
Superseded by https://github.com/ppy/osu/pull/1908
